### DISCUSSION TITLE
We see spikes in PostgreSQL connections since the Mastodon v3.5.4 upgrade

### DISCUSSION THREAD

**Discussion by marians** [mastodon/mastodon#20857] [General]
## Observation

In our instance we have had a rather steady number of Postgres connections, using much less than the configured limit of 100.

<img width="1344" alt="grafik" src="https://user-images.githubusercontent.com/273727/202126995-f5bd6867-6047-4e5e-a971-ccd0492c9737.png">

Since we have upgrade from Mastodon v3.5.3 to v3.5.4 and then v3.5.5 we have reached the connection several times, in short spikes.

<img width="1346" alt="grafik" src="https://user-images.githubusercontent.com/273727/202127327-ac570a07-7b68-4d5d-a458-40a27ed58c02.png">

In the above graph, the left vertical line marks the upgrade to Mastodon v3.5.4 and the right one marks the upgrade to v3.5.5.

What's not visible in the graph: the limit of 100 has been exceeded several times when trying to gather the connection metrics. we find error messages in our logs. If you look closely at the next graph, you can see these moments as interruptions in the graphs.

<img width="1339" alt="grafik" src="https://user-images.githubusercontent.com/273727/202127973-0efc9f90-19f3-4021-979c-937501205060.png">

## Speculation

What could be causing these spikes? The way we are collecting metrics has not changed over the time we are looking at. Has anything changed in Mastodon between v3.5.3 and v3.5.4 that could explain the difference? Potentially a dependency update related to the PostgreSQL client?

To mitigate, we have increased the connection limit far beyond the number usually needed. In terms of operational stability, this increases the risk that longer lasting connections will actually saturate the machine. Ideally we would find the cause for these spikes and prevent them.

We are planning to upgrade to Mastodon v4 tonight. I'll keep an eye on this graph.

**Comment by ralfr** [comment]
We saw a similar pattern and could link it back to more folks using third party mobile apps, which implement type-ahead search. Some implementations were so “bad” that they send every single character while a user is typing into the search field. This triggered the advanced search query hundreds of times per user per search. This query happens to be one of the more complex and long running.

Of course I don’t know whether this is the case in your scenario, too, but it might be worth monitoring the queries your db is executing while you’re seeing connection spikes.

**Comment by ineffyble** [comment]
You can see the changes in question here: https://github.com/mastodon/mastodon/compare/v3.5.3...v3.5.5

They were minimal, security backports. At a glance I can't see anything that would impact database connections.
